packages/admin-next/dashboard/src/v2-routes/campaigns/add-campaign-promotions/components/add-campaign-promotions-form.tsx

```diff
@@ -0,0 +1,225 @@
+import { useAddOrRemoveCampaignPromotions } from "../../../../hooks/api/campaigns"
+import { usePromotions } from "../../../../hooks/api/promotions"
+import { usePromotionTableColumns } from "../../../../hooks/table/columns-v2/use-promotion-table-columns"
+import { usePromotionTableFilters } from "../../../../hooks/table/filters/use-promotion-table-filters"
+import { usePromotionTableQuery } from "../../../../hooks/table/query-v2/use-promotion-table-query"
+import { useDataTable } from "../../../../hooks/use-data-table"
+
+type AddCampaignPromotionsFormProps = {
+  campaign: CampaignResponse
+}
+
+const AddCampaignPromotionsSchema = zod.object({
+  promotion_ids: zod.array(zod.string()).min(1),
+})
+
+const PAGE_SIZE = 50
+
+export const AddCampaignPromotionsForm = ({
+  campaign,
+}: AddCampaignPromotionsFormProps) => {
+  const { t } = useTranslation()
+  const { handleSuccess } = useRouteModal()
+
+  const form = useForm<zod.infer<typeof AddCampaignPromotionsSchema>>({
+    defaultValues: { promotion_ids: [] },
+    resolver: zodResolver(AddCampaignPromotionsSchema),
+  })
+
+  const { setValue } = form
+  const { mutateAsync, isPending } = useAddOrRemoveCampaignPromotions(
+    campaign.id
+  )
+  const [rowSelection, setRowSelection] = useState<RowSelectionState>({})
+
+  const updater: OnChangeFn<RowSelectionState> = (fn) => {
+    const state = typeof fn === "function" ? fn(rowSelection) : fn
+    const ids = Object.keys(state)
+
+    setValue("promotion_ids", ids, {
+      shouldDirty: true,
+      shouldTouch: true,
+    })
+
+    setRowSelection(state)
+  }
+
+  const { searchParams, raw } = usePromotionTableQuery({ pageSize: PAGE_SIZE })
+  const {
+    promotions,
+    count,
+    isPending: isLoading,
+    isError,
+    error,
+  } = usePromotions(
+    { ...searchParams, campaign_id: "null" },
+    { placeholderData: keepPreviousData }
+  )
+
+  const columns = useColumns()
+  const filters = usePromotionTableFilters()
+
+  const { table } = useDataTable({
+    data: promotions ?? [],
+    columns,
+    enableRowSelection: (row) => {
+      return row.original.campaign_id !== campaign.id
+    },
+    enablePagination: true,
+    getRowId: (row) => row.id,
+    pageSize: PAGE_SIZE,
+    count,
+    rowSelection: {
+      state: rowSelection,
+      updater,
+    },
+    meta: { campaignId: campaign.id },
+  })
+
+  const handleSubmit = form.handleSubmit(async (values) => {
+    await mutateAsync(
+      { add: values.promotion_ids },
+      {
+        onSuccess: () => {
+          toast.success(t("general.success"), {
+            description: t("campaigns.promotions.toast.success"),
+            dismissLabel: t("actions.close"),
+          })
+          handleSuccess()
+        },
+        onError: (error) =>
+          toast.error(t("general.error"), {
+            description: error.message,
+            dismissLabel: t("actions.close"),
+          }),
+      }
+    )
+  })
+
+  if (isError) {
+    throw error
+  }
```

Comment:
suggestion: I woudn't throw inside the form we only do this in a page container - we can just disaply no results UI or empty table